### DISCUSSION TITLE
Teslamate sends empty MQTT messages.

### DISCUSSION THREAD

**Discussion by Lizzard44** [teslamate-org/teslamate#3873] [General]
like "teslamate/ns2/cars/2/shift_state / (null)" ......

This was different until 1.28.5 and can confuse post-processing (openhab, Homeassistant, Arduino MQTT, etc.).

Wouldn't it be better not to send such messages in the first place?

**Comment by brianmay** [comment]
Sorry, I think you lost me.

There should not have been any changes to the behaviour of the `teslamate/ns2/cars/2/shift_state`. I think what you are talking is about is a regression that was in 1.28.5, that was not in 1.28.4.

Have I understood this correctly?

I believe the behaviour has always been to send an empty string - which can sometimes be displayed as "(null)" - when the car is not being driven.

We have to send something, because we set retain=true, and if we don't send something the old state will remain current. And it doesn't maker sense for a client to receive a shift state when the car is not being driven.

I don't think there were any changes to this code before 1.28.5, although I refactored the code in recent master (not yet released).

**Comment by Lizzard44** [comment]
Yes,

I searched my old logs
Previously "(null)" == 6 characters but now NULL == 0 characters is sent.
It looks similar in the log but it's something completely different.

It is permissible not to send anything. It's not Teslamate but everything else that needs to change.
I have already adjusted my program. ;)

This can be done here. It's been dealt with.
Thanks!

**Reply by brianmay** [reply]
> Previously "(null)" == 6 characters

Can you please confirm what version of teslamate did this? I have personally not seen anything like this myself.

**Comment by Lizzard44** [comment]
I can't read that from the log file. :(
I only noticed this because my C program from 1.28.5 onwards often crashed with a segfault.

> MyMQTT[5817]: segfault at 0 ip 00007f9289c61545 sp 00007ffdd35ff9a8 error 4 in libc.so.6[7f9289bd7000+154000]_

Of course that was my mistake because I didn't pay attention to the payloads and just continued working with the message pointer.

Then I noticed that Homeassistant (python) also reported errors more often.

> home-assistant.log: File "/usr/src/homeassistant/homeassistant/components/mqtt/client.py", line 964, in _raise_on_error_

I don't think it's a bug in Teslamate. Maybe something just changed in the elixir.
I haven't found anywhere that the message has to be filled in the MQTT-publish. 

only the Teslamate Doku writes
> teslamate/cars/$car_id/shift_state | D | Current/Last Shift State (D/N/R/P)
